### PR TITLE
dataflow/sink: properly indicate that the sink write frontier is empt…

### PR DESCRIPTION
**Pull request by ruchirK**
…y on shutdown.

Touches #8315

Previously, exactly-once sinks would update their write frontiers over time as
they wrote data, but could potentially end up in an error state where they
no longer update their write frontier, even though the sink dataflow itself is
not dropped.

In this case, the coordinator will hold onto timestamp bindings for all sources
the affected sink depends on indefinitely, which ultimately causes a resource
leak in sqlite.

### Checklist

- [ ] This PR has adequate test coverage.
- [ ] This PR adds a release note for any user-facing behavior changes.

### CODE REVIEW

**Comment by ruchirK** [comment]
Marking as a draft for now because this needs some more testing before I'll feel comfortable merging this

File: src/dataflow/src/sink/kafka.rs
Comment (ruchirK) [inline src/dataflow/src/sink/kafka.rs:708]: this is duplicative because theres only one path where we end up in `SendState::Shutdown` but I think 4 or so paths where we set the `shutdown_flag` to true

Comment (elindsey) [reply src/dataflow/src/sink/kafka.rs:708]: Does going into the shutdown state not cause this path to be hit on the next iteration?

Comment (ruchirK) [reply src/dataflow/src/sink/kafka.rs:708]: there's some paths that set the flag, but not the state, e.g. if we asynchronously discover that we failed to produce to the broker

```
 219     fn delivery(&self, result: &DeliveryResult, _: Self::DeliveryOpaque) {
 220         match result {
 221             Ok(_) => (),
 222             Err((e, msg)) => {
 223                 self.metrics.message_delivery_errors_counter.inc();
 224                 error!(
 225                     "received error while writing to kafka sink topic {}: {}",
 226                     msg.topic(),
 227                     e
 228                 );
 229                 self.shutdown_flag.store(true, Ordering::SeqCst);
 230             }
 231         }
 232     }
 ```

Comment (elindsey) [reply src/dataflow/src/sink/kafka.rs:708]: I meant the other way around - is there a problem with only clearing the write frontier here and not in the shutdown state?

Comment (ruchirK) [reply src/dataflow/src/sink/kafka.rs:708]: ah i see. yeah i think thats fine

Comment (elindsey) [reply src/dataflow/src/sink/kafka.rs:708]: Should I wait for that edit, or is this up for review in its current form?

Comment (ruchirK) [reply src/dataflow/src/sink/kafka.rs:708]: i made the edit! sorry just saw your question now

Comment (ruchirK) [reply src/dataflow/src/sink/kafka.rs:708]: oops - this was a brain fart where I thought I had made the edit but hadn't. actually pushed the edit now!

File: src/dataflow/src/sink/kafka.rs
Comment (aljoscha) [inline src/dataflow/src/sink/kafka.rs:912]: nit: we now log twice.

Comment (ruchirK) [reply src/dataflow/src/sink/kafka.rs:912]: yeah - i felt that was defensible as the other log site is a debug log and could be useful to know how many times we hit the sink operator even after its shutdown flag is set. Previously, we had an info log that could repeat indefinitely, so this seems better?

Comment (aljoscha) [reply src/dataflow/src/sink/kafka.rs:912]: Makes sense!

**Review by aljoscha — APPROVED** [review]
Now this looks very good!

**Pull request merged**
